sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/ExpressionSet.scala

```diff
@@ -74,49 +83,37 @@ class ExpressionSet protected(
 
   protected def remove(e: Expression): Unit = {
     if (e.deterministic) {
-      baseSet --= baseSet.filter(_ == e.canonicalized)
-      originals --= originals.filter(_.canonicalized == e.canonicalized)
+      baseSet.retain(_ != e.canonicalized)
+      originals = originals.filter(!_.semanticEquals(e))
```

Comment:
There are two changes happening in this line.
1. Changes to `originals = originals.filter(...)`. This is O(n) as the previous implementation is O(mn).
2. Uses `semanticEquals` instead of the previous `_.canonicalized == e.canonicalized` as the condition. `semanticEquals` can short circuit and avoid unnecessary access of `canonicalized` for non-deterministic expressions.